### ISSUE TITLE
Questions on Internals

### ISSUE THREAD

**Issue by polarathene** [gnif/LookingGlass#58]
I am presently working on a LG like client in Rust(I am more comfortable in this language than I am C++ and taking advantage of the third party libraries(crates)). Currently I am reading in frame updates from the LG server shared memory output, with no current plans for supporting more advanced formats like h264, nor SPICE(though I may try add that at a later date). After I will work on a server component that can support Linux VMs.

I have two questions so far:

1. You are setting a compositor bypass hint [here](https://github.com/gnif/LookingGlass/blob/fd8d4d3d383f59742c7db782c140e5a158fbd0a0/client/main.c#L795). [SDL sets this by default](https://github.com/ValveSoftware/Dota-2/issues/204#issuecomment-118540016), is it necessary? The specs also define it as [_NET_WM_BYPASS_COMPOSITOR](https://standards.freedesktop.org/wm-spec/wm-spec-latest.html#idm140200472547488), `xprop` can be used to verify this information. Your code is adding `NETWM_BYPASS_COMPOSITOR(CARDINAL) = 1`, is there some X11 implementation where they've implemented the spec incorrectly to read this property instead of the correct one, or was this a mistake on your part?

2. I'm not familiar enough with OpenGL to review the renderers yet. I've got a Client working that displays the frames by using SDLs canvas with updating a texture with each frame, copying to the canvas and presenting. I see you've gone a more direct/low-level route, how beneficial is that over the simple approach I'm presently going with?

**Comment by polarathene** [comment]
> NETWM_BYPASS_COMPOSITOR is required on some platforms because the compositor adds another frame redirection into the pipeline, we saw this on Ubuntu and Debian systems running Gnome Shell.

Just to confirm. You're stating `NETWM_BYPASS_COMPOSITOR` is required not `_NET_WM_BYPASS_COMPOSITOR` which is the correct value according to the spec? The latter which SDL automatically takes care of by default.

> Using the SDL canvas require an extra memory copy, performance suffers enormously.

I update the texture with the new pixel data(one copy) then the canvas requires a copy(2nd copy), that goes to the backbuffer and canvas present() then displays it(third copy? not sure how that part works). Performance seems pretty good for me currently, on an Intel i5-6500 host with iGPU frames take 2ms to complete. Perhaps it's more important for resolutions greater than 1080p

> The write locking on this is critical, the read locking isn't because it's a simple uint8_t.

I am not sure if an option of only write locking with Rust, I'll have to investigate. 

`state.running` is generally only read and in at least one place written to with false in a thread function. Is this not an atomic because it is rarely changed when the threading begins and it's not super important to care about timing/thread state being in sync for this bool?

> Flags isn't boolean, it is a bit set

Yes...I know that. I am stating that it's treated as a boolean for the listed conditionals. You implicitly get a true/false from your condition, whereas in Rust it's not permitted and I must either compare it(`uint8_t`/`u8`) against 0 or cast as a bool.

---

I will probably have to make some other changes to support threading. With your threads where you change several state fields(cursor and srcSize for example), although only one thread is doing such, Rust insists I handle it properly. I've moved cursor thread related fields into a cursorData struct and the same for frame thread fields in the state struct to a frameData struct(both structs are fields on the state struct). The new struct fields use a Mutex to lock them for writing(and apparently reading).

I made `state.running` an atomic, but this is probably not a good idea. The whole state struct also was wrapped in an Atomic Reference Counter(Arc) to be shared across threads(so state needs to use Mutex or Atomic to mutate).

Today I'll be looking at an alternative which uses channels instead. Rather than sharing the state from what I understand this will instead allow me to have a thread tell a thread that manages the state struct to update a field instead. The current approach probably isn't much better latency wise, Rust isn't going to be as laid back on C when dealing with threads though I think :(

Cheers for the feedback! :)

**Comment by polarathene** [comment]
`static inline void updatePositionInfo() {..}` This method is inlined into the 3 calls on `main.c`? I'm not using SDL threads and at least with Rust it appears that trying to deal with the window or anything rendering related on more than a single thread is ill advised(in my case it refuses to compile currently due to this).

Is there any harm in having `frameThread()` with it's two calls to the method instead set a bool(atomic) so that the main thread that runs the event pump(with the third call to the method in it) handles the logic instead? It doesn't look like that should cause any concerns in the behaviour of the logic? Another option could be to use thread channels rather than the bool to communicate to another thread to do the logic or request the width/height from the window while running the remaining function logic on frameThread() still... thoughts? I'll need to handle the params.autoResize condition too..

Your state and params structs are global right? I've created them on the main thread and passed them to the spawned threads. This caused a lifetime issue with state.shm which I have as a reference(not quite sure if I ported that correctly), I was able to resolve that by using scoped threads which allows for dropping the Arc usage, I've not done that yet, I don't think it affects perf for this code by much.
